File: scripts/grove-notify.swift
#!/usr/bin/env swift
// grove-notify.swift — Native macOS notification overlay for Grove Street
// Usage: grove-notify <sender> <phrase> <icon_path> <dismiss_seconds> <bundle_id> <project_name> <position> <slot_index> <slot_dir> [category_label]
//
// Positions: top-left, top-center, top-right, bottom-left, bottom-center, bottom-right, center

import Cocoa

// MARK: - Args

let args = CommandLine.arguments.dropFirst().map { $0 }
let senderName  = args.count > 0 ? args[0] : "Carl Johnson"
let phrase       = args.count > 1 ? args[1] : ""
let iconPath     = args.count > 2 ? args[2] : ""
let dismiss      = args.count > 3 ? Double(args[3]) ?? 4 : 4
let bundleId     = args.count > 4 ? args[4] : ""
let projectName  = args.count > 5 ? args[5] : "grove-street"
let position     = args.count > 6 ? args[6] : "top-right"
var slotIndex       = args.count > 7 ? Int(args[7]) ?? 0 : 0
let slotDir         = args.count > 8 ? args[8] : ""
let categoryLabel   = args.count > 9 ? args[9] : ""

let winWidth: CGFloat = 360
let winHeight: CGFloat = 68
let cornerR: CGFloat = 20
let margin: CGFloat = 12
let myPid = ProcessInfo.processInfo.processIdentifier

// MARK: - Slot management

func isSlotOccupied(_ idx: Int) -> Bool {
    guard !slotDir.isEmpty else { return true }
    let path = "\(slotDir)/\(idx).lock"
    guard FileManager.default.fileExists(atPath: path),
          let contents = try? String(contentsOfFile: path, encoding: .utf8) else {
        return false
    }
    guard let pid = Int32(contents.trimmingCharacters(in: .whitespacesAndNewlines)) else {
        return false
    }
    if pid == myPid { return true }
    return kill(pid, 0) == 0
}

func moveToSlot(_ newIdx: Int) {
    guard !slotDir.isEmpty else { return }
    let oldPath = "\(slotDir)/\(slotIndex).lock"
    let newPath = "\(slotDir)/\(newIdx).lock"
    try? "\(myPid)".write(toFile: newPath, atomically: true, encoding: .utf8)
    try? FileManager.default.removeItem(atPath: oldPath)
    slotIndex = newIdx
}

func cleanupSlot() {
    guard !slotDir.isEmpty else { return }
    try? FileManager.default.removeItem(atPath: "\(slotDir)/\(slotIndex).lock")
}

// MARK: - Position calculation

func calcOrigin(visibleFrame vf: NSRect, slot: Int) -> NSPoint {
    let stackOffset = CGFloat(slot) * (winHeight + margin)
    var x: CGFloat, y: CGFloat

    switch position {
    case "top-left":
        x = vf.origin.x + margin
        y = vf.origin.y + vf.size.height - winHeight - margin - stackOffset
    case "top-center":
        x = vf.origin.x + (vf.size.width - winWidth) / 2
        y = vf.origin.y + vf.size.height - winHeight - margin - stackOffset
    case "top-right":
        x = vf.origin.x + vf.size.width - winWidth - margin
        y = vf.origin.y + vf.size.height - winHeight - margin - stackOffset
    case "bottom-left":
        x = vf.origin.x + margin
        y = vf.origin.y + margin + stackOffset
    case "bottom-center":
        x = vf.origin.x + (vf.size.width - winWidth) / 2
        y = vf.origin.y + margin + stackOffset
    case "bottom-right":
        x = vf.origin.x + vf.size.width - winWidth - margin
        y = vf.origin.y + margin + stackOffset
    case "center":
        x = vf.origin.x + (vf.size.width - winWidth) / 2
        let halfOffset = stackOffset / 2
        y = vf.origin.y + (vf.size.height - winHeight) / 2 + (slot % 2 == 0 ? -halfOffset : halfOffset)
    default:
        x = vf.origin.x + vf.size.width - winWidth - margin
        y = vf.origin.y + vf.size.height - winHeight - margin - stackOffset
    }
    return NSPoint(x: x, y: y)
}

// MARK: - App setup

let app = NSApplication.shared
app.setActivationPolicy(.accessory)

var windows: [(window: NSWindow, screen: NSScreen)] = []

for screen in NSScreen.screens {
    let vf = screen.visibleFrame
    let origin = calcOrigin(visibleFrame: vf, slot: slotIndex)
    let frame = NSRect(x: origin.x, y: origin.y, width: winWidth, height: winHeight)

    let win = NSWindow(contentRect: frame, styleMask: .borderless, backing: .buffered, defer: false)
    win.backgroundColor = .clear
    win.isOpaque = false
    win.alphaValue = 0
    win.level = .statusBar
    win.collectionBehavior = [.canJoinAllSpaces, .stationary]
    win.hasShadow = true

    let contentView = win.contentView!
    contentView.wantsLayer = true

    // Liquid glass blur
    let effectView = NSVisualEffectView(frame: NSRect(x: 0, y: 0, width: winWidth, height: winHeight))
    effectView.material = .hudWindow
    effectView.blendingMode = .behindWindow
    effectView.state = .active
    effectView.alphaValue = 0.82
    effectView.wantsLayer = true
    effectView.layer?.cornerRadius = cornerR
    effectView.layer?.masksToBounds = true
    contentView.addSubview(effectView)

    // Glass edge border
    let borderBox = NSBox(frame: NSRect(x: 0, y: 0, width: winWidth, height: winHeight))
    borderBox.boxType = .custom
    borderBox.cornerRadius = cornerR
    borderBox.borderWidth = 0.5
    borderBox.borderColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.25)
    borderBox.fillColor = .clear
    borderBox.isTransparent = false
    effectView.addSubview(borderBox)

    // Top edge shine
    let edgeShine = NSBox(frame: NSRect(x: 30, y: winHeight - 1.5, width: winWidth - 60, height: 0.5))
    edgeShine.boxType = .custom
    edgeShine.cornerRadius = 0.25
    edgeShine.borderWidth = 0
    edgeShine.borderColor = .clear
    edgeShine.fillColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.18)
    edgeShine.isTransparent = false
    effectView.addSubview(edgeShine)

    var textX: CGFloat = 14
    var textWidth = winWidth - 28

    // Icon
    if !iconPath.isEmpty, FileManager.default.fileExists(atPath: iconPath),
       let iconImage = NSImage(contentsOfFile: iconPath) {
        let iconSz: CGFloat = 40
        let iconView = NSImageView(frame: NSRect(x: 14, y: (winHeight - iconSz) / 2, width: iconSz, height: iconSz))
        iconView.image = iconImage
        iconView.imageScaling = .scaleProportionallyUpOrDown
        iconView.wantsLayer = true
        iconView.layer?.cornerRadius = 10
        iconView.layer?.masksToBounds = true
        effectView.addSubview(iconView)
        textX = 14 + iconSz + 10
        textWidth = winWidth - textX - 14
    }

    // Text layout
    let lineHeight1: CGFloat = 18
    let lineHeight2: CGFloat = 16
    let gap: CGFloat = 2
    let totalHeight = lineHeight1 + gap + lineHeight2
    let baseY = (winHeight - totalHeight) / 2

    // Line 1 left: sender in project
    let senderLabel = NSTextField(frame: NSRect(x: textX, y: baseY + lineHeight2 + gap, width: textWidth, height: lineHeight1))
    senderLabel.stringValue = "\(senderName) in \(projectName)"
    senderLabel.isBezeled = false
    senderLabel.drawsBackground = false
    senderLabel.isEditable = false
    senderLabel.isSelectable = false
    senderLabel.textColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.95)
    senderLabel.font = .boldSystemFont(ofSize: 13.5)
    senderLabel.lineBreakMode = .byTruncatingTail
    effectView.addSubview(senderLabel)

    // Line 1 right: category label (vertically centered within the sender line)
    if !categoryLabel.isEmpty {
        let catFont = NSFont.systemFont(ofSize: 10.5)
        let catH: CGFloat = 14
        let catY = baseY + lineHeight2 + gap + (lineHeight1 - catH) / 2
        let catLabel = NSTextField(frame: NSRect(x: textX, y: catY, width: textWidth, height: catH))
        catLabel.stringValue = categoryLabel
        catLabel.isBezeled = false
        catLabel.drawsBackground = false
        catLabel.isEditable = false
        catLabel.isSelectable = false
        catLabel.textColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.6)
        catLabel.font = catFont
        catLabel.alignment = .right
        catLabel.lineBreakMode = .byClipping
        effectView.addSubview(catLabel)
    }

    // Line 2: phrase
    if !phrase.isEmpty {
        let msgLabel = NSTextField(frame: NSRect(x: textX, y: baseY, width: textWidth, height: lineHeight2))
        msgLabel.stringValue = phrase
        msgLabel.isBezeled = false
        msgLabel.drawsBackground = false
        msgLabel.isEditable = false
        msgLabel.isSelectable = false
        msgLabel.textColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.55)
        msgLabel.font = .systemFont(ofSize: 12)
        msgLabel.lineBreakMode = .byTruncatingTail
        msgLabel.cell?.wraps = false
        effectView.addSubview(msgLabel)
    }

    // Click button (transparent look, but NOT .isTransparent which disables hit-testing)
    let btn = NSButton(frame: NSRect(x: 0, y: 0, width: winWidth, height: winHeight))
    btn.title = ""
    btn.isBordered = false
    btn.isTransparent = false
    btn.alphaValue = 0.001
    btn.target = nil
    btn.action = #selector(NSApplication.terminate(_:))
    effectView.addSubview(btn)

    win.orderFrontRegardless()
    NSAnimationContext.runAnimationGroup { ctx in
        ctx.duration = 0.2
        win.animator().alphaValue = 1.0
    }

    windows.append((window: win, screen: screen))
}

// MARK: - Click handler to focus app

class AppDelegate: NSObject, NSApplicationDelegate {
    func applicationWillTerminate(_ notification: Notification) {
        if !bundleId.isEmpty {
            if let targetApp = NSRunningApplication.runningApplications(withBundleIdentifier: bundleId).first {
                targetApp.activate()
            }
        }
        cleanupSlot()
    }
}

let delegate = AppDelegate()
app.delegate = delegate

// MARK: - Reflow timer

if !slotDir.isEmpty {
    Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { _ in
        for s in 0..<slotIndex {
            if !isSlotOccupied(s) {
                moveToSlot(s)
                NSAnimationContext.runAnimationGroup { ctx in
                    ctx.duration = 0.25
                    ctx.allowsImplicitAnimation = true
                    for entry in windows {
                        let newOrigin = calcOrigin(visibleFrame: entry.screen.visibleFrame, slot: slotIndex)
                        var newFrame = entry.window.frame
                        newFrame.origin = newOrigin
                        entry.window.animator().setFrame(newFrame, display: true)
                    }
                }
                break
            }
        }
    }
}

// MARK: - Auto-dismiss

if dismiss > 0 {
    Timer.scheduledTimer(withTimeInterval: dismiss, repeats: false) { _ in
        NSAnimationContext.runAnimationGroup({ ctx in
            ctx.duration = 0.3
            for entry in windows {
                entry.window.animator().alphaValue = 0
            }
        }, completionHandler: {
            app.terminate(nil)
        })
    }
}

app.run()

File: scripts/grove-notify.py
#!/usr/bin/env python3
"""grove-notify.py — Native Linux notification overlay for Grove Street.

Usage: grove-notify.py <sender> <phrase> <icon_path> <dismiss_seconds> <bundle_id> <project_name> <position> <slot_index> <slot_dir>

Positions: top-left, top-center, top-right, bottom-left, bottom-center, bottom-right, center
"""

import os
import sys
import signal

import gi
gi.require_version("Gtk", "3.0")
gi.require_version("Gdk", "3.0")
from gi.repository import Gtk, Gdk, GLib, GdkPixbuf, Pango

# --- Args ---

args = sys.argv[1:]
sender_name = args[0] if len(args) > 0 else "Carl Johnson"
phrase = args[1] if len(args) > 1 else ""
icon_path = args[2] if len(args) > 2 else ""
dismiss_secs = float(args[3]) if len(args) > 3 else 4.0
bundle_id = args[4] if len(args) > 4 else ""  # unused on Linux
project_name = args[5] if len(args) > 5 else "grove-street"
position = args[6] if len(args) > 6 else "top-right"
slot_index = int(args[7]) if len(args) > 7 else 0
slot_dir = args[8] if len(args) > 8 else ""

WIN_WIDTH = 360
WIN_HEIGHT = 68
CORNER_R = 16
MARGIN = 12
MY_PID = os.getpid()

# --- Slot management ---


def is_slot_occupied(idx):
    if not slot_dir:
        return True
    path = os.path.join(slot_dir, f"{idx}.lock")
    if not os.path.exists(path):
        return False
    try:
        pid = int(open(path).read().strip())
    except (ValueError, OSError):
        return False
    if pid == MY_PID:
        return True
    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


def move_to_slot(new_idx):
    global slot_index
    if not slot_dir:
        return
    old_path = os.path.join(slot_dir, f"{slot_index}.lock")
    new_path = os.path.join(slot_dir, f"{new_idx}.lock")
    try:
        with open(new_path, "w") as f:
            f.write(str(MY_PID))
        os.remove(old_path)
    except OSError:
        pass
    slot_index = new_idx


def cleanup_slot():
    if not slot_dir:
        return
    path = os.path.join(slot_dir, f"{slot_index}.lock")
    try:
        os.remove(path)
    except OSError:
        pass


# --- Position calculation ---


def calc_origin(screen_geo, slot):
    """Returns (x, y) for the notification window."""
    sx, sy, sw, sh = screen_geo
    stack_offset = slot * (WIN_HEIGHT + MARGIN)

    if position == "top-left":
        x = sx + MARGIN
        y = sy + MARGIN + stack_offset
    elif position == "top-center":
        x = sx + (sw - WIN_WIDTH) // 2
        y = sy + MARGIN + stack_offset
    elif position == "top-right":
        x = sx + sw - WIN_WIDTH - MARGIN
        y = sy + MARGIN + stack_offset
    elif position == "bottom-left":
        x = sx + MARGIN
        y = sy + sh - WIN_HEIGHT - MARGIN - stack_offset
    elif position == "bottom-center":
        x = sx + (sw - WIN_WIDTH) // 2
        y = sy + sh - WIN_HEIGHT - MARGIN - stack_offset
    elif position == "bottom-right":
        x = sx + sw - WIN_WIDTH - MARGIN
        y = sy + sh - WIN_HEIGHT - MARGIN - stack_offset
    elif position == "center":
        x = sx + (sw - WIN_WIDTH) // 2
        half = stack_offset // 2
        y = sx + (sh - WIN_HEIGHT) // 2 + (-half if slot % 2 == 0 else half)
    else:  # default top-right
        x = sx + sw - WIN_WIDTH - MARGIN
        y = sy + MARGIN + stack_offset

    return x, y


# --- CSS ---

CSS = b"""
window {
    background-color: transparent;
}

#notification-box {
    background-color: rgba(30, 30, 30, 0.88);
    border-radius: 16px;
    border: 1px solid rgba(255, 255, 255, 0.12);
    padding: 12px 14px;
}

#sender-label {
    color: rgba(255, 255, 255, 0.92);
    font-weight: bold;
    font-size: 13px;
}

#phrase-label {
    color: rgba(255, 255, 255, 0.50);
    font-size: 12px;
}
"""

# --- Window ---


class NotificationWindow(Gtk.Window):
    def __init__(self):
        super().__init__(type=Gtk.WindowType.POPUP)

        self.set_decorated(False)
        self.set_resizable(False)
        self.set_default_size(WIN_WIDTH, WIN_HEIGHT)
        self.set_keep_above(True)
        self.set_skip_taskbar_hint(True)
        self.set_skip_pager_hint(True)
        self.stick()

        # Transparency
        screen = self.get_screen()
        visual = screen.get_rgba_visual()
        if visual:
            self.set_visual(visual)
        self.set_app_paintable(True)

        # Apply CSS
        css_provider = Gtk.CssProvider()
        css_provider.load_from_data(CSS)
        Gtk.StyleContext.add_provider_for_screen(
            screen, css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION
        )

        # Main container
        event_box = Gtk.EventBox()
        event_box.connect("button-press-event", self._on_click)
        self.add(event_box)

        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=10)
        box.set_name("notification-box")
        event_box.add(box)

        # Icon
        if icon_path and os.path.exists(icon_path):
            try:
                pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_scale(
                    icon_path, 40, 40, True
                )
                icon = Gtk.Image.new_from_pixbuf(pixbuf)
                box.pack_start(icon, False, False, 0)
            except GLib.Error:
                pass

        # Text
        text_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
        text_box.set_valign(Gtk.Align.CENTER)
        box.pack_start(text_box, True, True, 0)

        sender_label = Gtk.Label()
        sender_label.set_name("sender-label")
        sender_label.set_markup(
            f"<b>{GLib.markup_escape_text(sender_name)} in {GLib.markup_escape_text(project_name)}</b>"
        )
        sender_label.set_xalign(0)
        sender_label.set_ellipsize(Pango.EllipsizeMode.END)
        text_box.pack_start(sender_label, False, False, 0)

        if phrase:
            phrase_label = Gtk.Label(label=phrase)
            phrase_label.set_name("phrase-label")
            phrase_label.set_xalign(0)
            phrase_label.set_ellipsize(Pango.EllipsizeMode.END)
            text_box.pack_start(phrase_label, False, False, 0)

        self._screen_geo = None
        self._position_window()

        # Reflow timer
        if slot_dir:
            GLib.timeout_add(500, self._check_reflow)

        # Auto-dismiss
        if dismiss_secs > 0:
            GLib.timeout_add(int(dismiss_secs * 1000), self._dismiss)

    def _get_screen_geo(self):
        display = Gdk.Display.get_default()
        monitor = display.get_primary_monitor() or display.get_monitor(0)
        if monitor:
            geo = monitor.get_workarea()
            return (geo.x, geo.y, geo.width, geo.height)
        # Fallback
        screen = self.get_screen()
        return (0, 0, screen.get_width(), screen.get_height())

    def _position_window(self):
        self._screen_geo = self._get_screen_geo()
        x, y = calc_origin(self._screen_geo, slot_index)
        self.move(x, y)

    def _check_reflow(self):
        global slot_index
        for s in range(slot_index):
            if not is_slot_occupied(s):
                move_to_slot(s)
                self._position_window()
                break
        return True  # keep timer running

    def _on_click(self, widget, event):
        # Try to focus parent app using wmctrl or xdotool
        _try_focus_parent()
        self._dismiss()

    def _dismiss(self):
        cleanup_slot()
        Gtk.main_quit()
        return False


def _try_focus_parent():
    """Best-effort focus of the terminal/IDE that spawned us."""
    ppid = os.getppid()
    # Try xdotool first
    os.system(f"xdotool search --pid {ppid} windowactivate 2>/dev/null")


def main():
    win = NotificationWindow()
    win.show_all()
    Gtk.main()


if __name__ == "__main__":
    main()
